tools/checklocks/state.go

```diff
@@ -350,26 +391,21 @@ func (l *lockState) String() string {
 			locksHeld = append(locksHeld, k)
 		}
 	}
-	// sort for comparing results 
+	// sort for comparing results
 	sort.Strings(locksHeld)
 
 	return strings.Join(locksHeld, ",")
 }
 
-// pushDefer pushes a defer onto the stack.
-func (l *lockState) pushDefer(d *ssa.Defer) {
-	l.modify()
-	l.defers = append(l.defers, d)
-}
-
-// popDefer pops a defer from the stack.
-func (l *lockState) popDefer() *ssa.Defer {
-	// Does not technically modify the underlying slice.
-	count := len(l.defers)
-	if count == 0 {
-		return nil
+func (l *lockState) resolveDeferredUnlocks() (string, bool) {
+	for k, lockType := range l.mutexes {
+		if lockType.is(locked.union(deferredunlock)) {
+			l.modify()
+			l.mutexes[k] = unlocked
+		} else if lockType.is(deferredunlock) {
+			return k, false
```

Comment:
It seems odd to stop the loop early with some modifications to l.mutexes but not all of them. Is this intentional? If so, this might need to be documented.